README.md
# PostgreSQL STIG Compliance Validator

PostgreSQL STIG Compliance Validator (pgStigCheck) for InSpec is an open source compliance testing solution for PostgreSQL.

Developed in order to reduce the time it takes to secure Authority to Operate (ATO) certification for cloud services, pgStickCheck technology leverages open source software to provide automated compliance testing in real time. pgStigCheck uses the [InSpec](https://github.com/chef/inspec), which provides an open source compliance, security and policy testing framework that dynamically extracts system configuration information.

## PostgreSQL STIG Overview

The [PostgreSQL Security Technical Implementation Guide](https://www.crunchydata.com/postgres-stig/PGSQL-STIG-9.5+.pdf) (STIG) by the United States Defense Information Systems Agency (DISA) offers security-conscious enterprises a comprehensive guide for the configuration and operation of open source PostgreSQL.

[STIG](https://en.wikipedia.org/wiki/Security_Technical_Implementation_Guide)s are the configuration standards for United States Department of Defense (DoD) Information Assurance (IA) and IA-enabled devices/systems published by the United States Defense Information Systems Agency (DISA). Since 1998, DISA has played a critical role enhancing the security posture of DoD's security systems by providing the STIGs. The STIGs contain technical guidance to "lock down" information systems/software that might otherwise be vulnerable to a malicious computer attack.

The requirements associated with the PostgreSQL STIG are derived from the [National Institute of Standards and Technology](https://en.wikipedia.org/wiki/National_Institute_of_Standards_and_Technology) (NIST) [Special Publication (SP) 800-53, Revision 4](https://en.wikipedia.org/wiki/NIST_Special_Publication_800-53) and related documents.

While the PostgreSQL STIG was developed to provide technical guidance to "lock down" information systems used within the DoD, the guidance applies to all organizations that need to meet internal security as well as compliance standards.

The PostgreSQL STIG provides guidance on the configuration of PostgreSQL to address requirements associated with:

- Authentication
- Access Controls
- Data encryption at rest and over the wire
- Auditing
- Logging
- Administration
- Protection against SQL Injection

## Getting Started

### Requirements

To run the PostgreSQL STIG Compliance Validator, there are specific requirements on both the database host as will as the STIG valudation host.

#### Database Host

- PostgreSQL 9.5+ cluster running on \*nix host
- Remote access to PostgreSQL Server
- lsof
- netstat

#### STIG Validation Execution Host

- Linux VM or Host
- sudo access to install packages

#### Required software on STIG Validation Execution Host

- git
- ssh
- [InSpec](https://www.chef.io/products/chef-inspec/)

### Setup Environment on STIG Validation Execution Host

#### Install InSpec via your System Package Manager (recommended)

The InSpec community and chef provide packages for all major platforms. The installation package will bring all needed libraries and components needed by InSpec.

This is recommended for production and LTS environments.

Goto <https://downloads.chef.io/inspec/stable> and copy download link

For example:

- <https://packages.chef.io/files/stable/inspec/3.9.3/el/7/inspec-3.9.3-1.el7.x86_64.rpm>

```sh
sudo yum insall https://packages.chef.io/files/stable/inspec/3.9.3/el/7/inspec-3.9.3-1.el7.x86_64.rpm
```

#### Ensure InSpec Installed and is 3.x or higher

```sh
inspec --version
```

#### Install InSpec via GEM and RVM (alternative)

If you already have an existing Ruby environment configured on your system, or use RVM to manage your Ruby environments, you can always just install the InSpec gem and its dependencies using GEM.

```sh
$ curl -sSL https://rvm.io/mpapis.asc | gpg2 --import -
$ curl -sSL https://rvm.io/pkuczynski.asc | gpg2 --import -
$ curl -L get.rvm.io | bash -s stable
$ rvm install 2.4.0
$ rvm use 2.4.0 --default

$ source ${HOME}/.rvm/scripts/rvm
```

#### Install the needed gems

```sh
gem install inspec
```

#### Ensure InSpec Installed and is 3.x or higher

```sh
inspec --version
```

### Setting & Reviewing the Core Profile Attributes

The `core` or `shared` attributes are set in the `inspec.yml` file in the profile. This stores the default values for the
shared attributes that the profile uses to examine your system.

These attributes _should not be edited directly_!

If you need to override the default values for the core attributes to better match your system under evaluation, please:

1. Review the attributes and their defaults in the `inspec.yml` and note the attributes/inputs you need to tailor for your installation.
2. Create a `system` or `instance` specific `attributes.yml` - such as `attributes.mysystem.yml` and override the attributes with this file.
3. When you run the profile, you can load your updated attributes using the `--attrs` flag on the `inspec exec` command (see below)

### Sensitive Attributes (aka `passwords`, `system accounts` and `owners` in your PostgreSQL database )

The recommend way to store sensitive information is to use one of the environmental variables outlined below. For instance, you can set the password for the PostgreSQL user with the `PG_USER_PWD` environmental variable.

Alternatively, you can set these variables in your `attributes.mysystem.yml` file, but again, this is not recommended for sensitive information like passwords.

#### Set the following `environment variables` before you run the profile:

- PG_OWNER
- PG_OWNER_GRP
- PG_OWNER_PWD
- PG_DBA
- PG_DBA_PWD
- PG_USER
- PG_USER_PWD
- PG_HOST
- PG_PORT
- LOGIN_USER
- LOGIN_HOST
- PG_SYSLOG_OWNER

### Examples for your `attributes.mysystem.yml`

#### OS Group, User

```yaml
pg_owner: "postgres"
pg_group: "postgres"
# password set via `env_var`
```

#### DBA User

```yaml
pg_dba: "dba"
# password set via `env_var`
```

#### Normal DB User

```yaml
pg_user: "<username>"
# password set via `env_var`
```

#### DB Host and Port

```yaml
pg_host: "127.0.0.1"
pg_port: "5432"
```

#### DB Name and Test table

```yaml
pg_db: "test_db"
pg_table: "test_table"
```

#### Misc settings

```yaml
login_user: "<user on remote DB server>"
login_host: "<DB Host IP>"
pg_version: "9.5"
pg_data_dir: "/var/lib/pgsql/9.5/data"
pg_conf_file: "/var/lib/pgsql/9.5/data/postgresql.conf"
pg_user_defined_conf: "/var/lib/pgsql/9.5/data/stig-postgresql.conf"
pg_hba_conf_file: "/var/lib/pgsql/9.5/data/pg_hba.conf"
pg_ident_conf_file: "/var/lib/pgsql/9.5/data/pg_ident.conf"
pg_shared_dirs:
  [
    "/usr/pgsql-9.5",
    "/usr/pgsql-9.5/bin",
    "/usr/pgsql-9.5/lib",
    "/usr/pgsql-9.5/share",
  ]
pg_conf_mode: "0600"
pg_ssl: "on"
pg_log_dest: "syslog"
pg_syslog_facility: ["local0"]
pg_syslog_owner: "postgres"
pgaudit_log_items: ["ddl", "role", "read", "write"]
pgaudit_log_line_items: ["%m", "%u", "%c"]
pg_superusers: ["postgres"]
pg_users: []
pg_replicas: ["192.168.1.3/32"]
pg_max_connections: "100"
pg_timezone: "UTC"
```

### Validating Your PostgreSQL Instance

(See: <https://www.inspec.io/docs/reference/cli/>)

#### Execute a single Control in the Profile

**Note**: replace the profile's directory name - e.g. - `pgstigcheck-inspec` with `.` if you are in the profile's root directory.

```sh
inspec exec pgstigcheck-inspec/controls/V-72845.rb --attrs attributes.mysystem.yml -i <your ssh private key> -t ssh://<user>@<db host>:<port> --reporter cli json:myresults.json
```

or use the `--controls` flag

```sh
inspec exec pgstigcheck-inspec --controls=V-72845 V-72861 --attrs attributes.mysystem.yml  -i <your ssh private key>  -t ssh://<user>@<db host>:<port> --reporter cli json:myresults.json
```

#### Execute a Single Control and save results as HTML

```sh
inspec exec pgstigcheck-inspec --controls=V-72845 --attrs attributes.mysystem.yml -i <your ssh private key> --sudo --sudo-options="-u postgres" -t ssh://<user>@<db host>:<port> --reporter cli html:myresults.html
```

#### Execute All Controls in the Profile

```sh
inspec exec pgstigcheck-inspec --attrs attributes.yml -i <your ssh private key> --sudo --sudo-options="-u postgres"  -t ssh://<user>@<db host>:<port> --reporter cli json:myresults.json
```

#### Execute all the Controls in the Profile and save results as HTML

```sh
inspec exec pgstigcheck-inspec --attrs attributes.yml -i <your ssh private key> --sudo --sudo-options="-u postgres" -t ssh://<user>@<db host>:<port> --reporter cli html:myresults.html
```

### Reviewing your results

You can review your results from above in many ways, as you saw your results came back in multiple outputs - on the cli and in either `json` or `html`.

You can learn more about the different [InSpec Reporters](https://www.inspec.io/docs/reference/reporters/) on the [inspec.io](http://www.inspec.io) site.

## Pro Tips

### Location, Location, Location

The `--reporters` flags **must _always_** be at the end of your `inspec exec` cli command as they can user either `=` or `spaces` and so they must be at the end of the command.

If you used the examples above, you should have a `myresults.json` or `myresults.html` which you can review.

### HTML Reporter

The `myresults.html` in our examples - aka the InSpec HTML Reporter - is a working `html` file report but its output is very `technical` and is not recommended for **security review** or **accreditation discussions**.

### Use MITRE Hiemdall or Heimdall-Lite

Use the `JSON` InSpec Reporter output and the MITRE [Heimdall-Lite](http://mitre.github.io/heimdall-lite) for the best possible view of the results.

The **recommended** review format for for **security review** or **accreditation discussions** is the `JSON` results format using the InSpec `JSON` reporter and the MITRE `heimdall-lite` viewer.

You can use heimdall-lite any-time anywhere from: <http://mitre.github.io/heimdall-lite/>. Heimdall-Lite is a Single Page Client Side JavaScript app that runs completely in your browser and was designed to help make reviewing, sorting and sharing your InSpec results easier.

You can also download the `.html` files via a simple `save as` from your browser should you need to use `heimdall-lite` in a disconnected setting.

#### Heimdall vs Heimdall-Lite

[Heimdall-Lite](http://mitre.github.io/heimdall-lite) is a [VueJS](https://vuejs.org/) powered client side only view of your data for teams and devs doing their security compliance work.

If you need a more ongoing compliance view of your InSpec results, get the full [MITRE Hiemdall](https://www.github.com/mitre/heimdall) application / server which provides enhanced capabilities - like storage, timelines and more - and is powered by Rails and CrunchyDB PostgreSQL.

You can find out more about the InSpec Tools and Open Source applications at <http://inspec.mitre.org>.

## Sponsors

[![Crunchy Data](/hugo/static/images/crunchy_logo.png)](https://www.crunchydata.com/)

[Crunchy Data](https://www.crunchydata.com/) is pleased to sponsor pgstigcheck-inspec and many other [open-source projects](https://github.com/CrunchyData/) to help promote support the PostgreSQL community and software ecosystem.

[![The MITRE Corporation](https://upload.wikimedia.org/wikipedia/commons/d/d3/Mitre_Corporation_logo.svg)](https://www.mitre.org/)

[The MITRE Corporation](https://www.mitre.org) is pleased to support our Sponsors and CrunchyData in the creation of the PostgreSQL 9.x STIG and the pgstigcheck-inspec validation profile. MITRE also supports many other [inspec validation baselines](https://github.com/mitre?&q=baseline) on the MITRE GitHub (https://github.com/mitre/) in the [Public Interest](https://www.mitre.org/about/mission-and-values).

---

## Legal Notices

Copyright © 2019 Crunchy Data Solutions, Inc.

CRUNCHY DATA SOLUTIONS, INC. PROVIDES THIS GUIDE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF NON INFRINGEMENT, MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.

Crunchy, Crunchy Data Solutions, Inc. and the Crunchy Hippo Logo are trademarks of Crunchy Data Solutions, Inc.


CHANGELOG.md
# Changelog
All notable changes to crunchy-hugo-theme will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Develop]

### Added

### Changed

### Removed

### Fixed

## [2.2.0] - 2019-02-23

### Removed

- Removed references to a Highlight JavaScript file which was not in use.

### Fixed

- The chevron FontAwesome icons on the side navigation menu for the crunchy-containers and postgres-operator projects were originally "not found" on page load of specifically parent pages with children, and the icons would not show as expected. They now load as expected.

- A jQuery scroll fix was added; when clicking on an anchor, the header for that section no longer hides behind the top sticky navigation menu.


controls/V-72843.rb
# encoding: utf-8
#
=begin
-----------------
Benchmark: PostgreSQL 9.x Security Technical Implementation Guide
Status: Accepted

This Security Technical Implementation Guide is published as a tool to improve
the security of Department of Defense (DoD) information systems. The
requirements are derived from the National Institute of Standards and
Technology (NIST) 800-53 and related documents. Comments or proposed revisions
to this document should be sent via email to the following address:
disa.stig_spt@mail.mil.

Release Date: 2017-01-20
Version: 1
Publisher: DISA
Source: STIG.DOD.MIL
uri: http://iase.disa.mil
-----------------
=end

PG_CONF_FILE= attribute(
  'pg_conf_file',
  description: 'define path for the postgresql configuration file',
  default: '/var/lib/pgsql/9.5/data/postgresql.conf'
)

PG_DBA = attribute(
  'pg_dba',
  description: 'The postgres DBA user to access the test database',
)

PG_DBA_PASSWORD = attribute(
  'pg_dba_password',
  description: 'The password for the postgres DBA user',
)

PG_DB = attribute(
  'pg_db',
  description: 'The database used for tests',
)

PG_HOST = attribute(
  'pg_host',
  description: 'The hostname or IP address used to connect to the database',
)


  # @todo we need to provide a better example of the error_log output
  control "V-72843" do
  title "PostgreSQL must produce audit records containing sufficient information
  to establish the outcome (success or failure) of the events."
  desc  "Information system auditing capability is critical for accurate
  forensic analysis. Without information about the outcome of events, security
  personnel cannot make an accurate assessment as to whether an attack was
  successful or if changes were made to the security state of the system.
  Event outcomes can include indicators of event success or failure and
  event-specific results (e.g., the security state of the information system
  after the event occurred). As such, they also provide a means to measure the
  impact of an event and help authorized personnel to determine the appropriate
  response."
  impact 0.5
  tag "severity": "medium"
  tag "gtitle": "SRG-APP-000099-DB-000043"
  tag "gid": "V-72843"
  tag "rid": "SV-87495r1_rule"
  tag "stig_id": "PGS9-00-000200"
  tag "cci": "CCI-000134"
  tag "nist": ["AU-3", "Rev_4"]
  tag "check": "Note: The following instructions use the PGDATA environment
  variable. See supplementary content APPENDIX-F for instructions on configuring
  PGDATA.

    As a database administrator (shown here as \"postgres\"), create a table,
    insert a value, alter the table and update the table by running the following
    SQL:

    CREATE TABLE stig_test_table(id INT);
    INSERT INTO stig_test_table(id) VALUES (0);
    ALTER TABLE stig_test_table ADD COLUMN name text;
    UPDATE stig_test_table SET id = 1 WHERE id = 0;

    Next, as a user without access to the stig_test table, run the following SQL:

    INSERT INTO stig_test_table(id) VALUES (1);
    ALTER TABLE stig_test_table DROP COLUMN name;
    UPDATE stig_test_table SET id = 0 WHERE id = 1;

    The prior SQL should generate errors:

    ERROR: permission denied for relation stig_test
    ERROR: must be owner of relation stig_test
    ERROR: permission denied for relation stig_test

    Now, as the database administrator, drop the test table by running the
    following SQL:

    DROP TABLE stig_test_table;

    Now verify the errors were logged:

    $ sudo su - postgres
    $ cat ${PGDATA?}/pg_log/<latest_logfile>$PGDATA/
    < 2016-02-23 14:51:31.103 EDT psql postgres postgres 570bf22a.3af2 2016-04-11
    14:51:22 EDT [local] >LOG: AUDIT: SESSION,1,1,DDL,CREATE TABLE,,,CREATE TABLE
    stig_test(id INT);,<none> < 2016-02-23 14:51:44.835 EDT psql postgres
    postgres 570bf22a.3af2 2016-04-11 14:51:22 EDT [local] >LOG: AUDIT:
    SESSION,2,1,WRITE,INSERT,,,INSERT INTO stig_test(id) VALUES (0);,<none>
    < 2016-02-23 14:53:25.805 EDT psql postgres postgres 570bf22a.3af2 2016-04-11
    14:51:22 EDT [local] >LOG: AUDIT: SESSION,3,1,DDL,ALTER TABLE,,,ALTER TABLE
    stig_test ADD COLUMN name text;,<none> < 2016-02-23 14:53:54.381 EDT psql
    postgres postgres 570bf22a.3af2 2016-04-11 14:51:22 EDT [local] >LOG: AUDIT:
    SESSION,4,1,WRITE,UPDATE,,,UPDATE stig_test SET id = 1 WHERE id = 0;,<none>
    < 2016-02-23 14:54:20.832 EDT psql postgres postgres 570bf22a.3af2 2016-04-11
    14:51:22 EDT [local] >ERROR: permission denied for relation stig_test
    < 2016-02-23 14:54:20.832 EDT psql postgres postgres 570bf22a.3af2 2016-04-11
    14:51:22 EDT [local] >STATEMENT: INSERT INTO stig_test(id) VALUES (1);
    < 2016-02-23 14:54:41.032 EDT psql postgres postgres 570bf22a.3af2 2016-04-11
    14:51:22 EDT [local] >ERROR: must be owner of relation stig_test < 2016-02-23
    14:54:41.032 EDT psql postgres postgres 570bf22a.3af2 2016-04-11 14:51:22 EDT
    [local] >STATEMENT: ALTER TABLE stig_test DROP COLUMN name; < 2016-02-23
    14:54:54.378 EDT psql postgres postgres 570bf22a.3af2 2016-04-11 14:51:22 EDT
    [local] >ERROR: permission denied for relation stig_test < 2016-02-23
    14:54:54.378 EDT psql postgres postgres 570bf22a.3af2 2016-04-11 14:51:22 EDT
    [local] >STATEMENT: UPDATE stig_test SET id = 0 WHERE id = 1; < 2016-02-23
    14:55:23.723 EDT psql postgres postgres 570bf307.3b0a 2016-04-11 14:55:03 EDT
    [local] >LOG: AUDIT: SESSION,1,1,DDL,DROP TABLE,,,DROP TABLE stig_test;,<none>

    If audit records exist without the outcome of the event that occurred, this is
    a finding."

    tag "fix": "Using pgaudit PostgreSQL can be configured to audit various facets
    of PostgreSQL. See supplementary content APPENDIX-B for documentation on
    installing pgaudit.
    All errors, denials and unsuccessful requests are logged if logging is enabled.
    See supplementary content APPENDIX-C for documentation on enabling logging.
    Note: The following instructions use the PGDATA environment variable. See
    supplementary content APPENDIX-F for instructions on configuring PGDATA.
    With pgaudit and logging enabled, set the following configuration settings in
    postgresql.conf, as the database administrator (shown here as \"postgres\"),
    to the following:

    $ vi ${PGDATA?}/postgresql.conf
    pgaudit.log_catalog='on'
    pgaudit.log_level='log'
    pgaudit.log_parameter='on'
    pgaudit.log_statement_once='off'
    pgaudit.log='all, -misc'

    Next, tune the following logging configurations in postgresql.conf:

    $ sudo vi ${PGDATA?}/postgresql.conf
    log_line_prefix = '%m %u %d %e : '
    log_error_verbosity = default

    Last, as the system administrator, restart PostgreSQL:

    # SERVER USING SYSTEMCTL ONLY
    $ sudo systemctl restart postgresql-9.5

    # SERVER USING INITD ONLY
    $ sudo service postgresql-9.5 restart"

  # @FIXME related to Issue#1

    admin_sql = postgres_session(PG_DBA, PG_DBA_PASSWORD, PG_HOST)

  #   these are the set-up actions for test
    

    describe admin_sql.query('DROP TABLE IF EXISTS stig_test;', [PG_DB]) do
      its('output') { should eq 'DROP TABLE' }
    end

    describe admin_sql.query('CREATE TABLE stig_test(id INT);', [PG_DB]) do
      its('output') { should eq 'CREATE TABLE' }
    end

    describe admin_sql.query('INSERT INTO stig_test(id) VALUES (0);', [PG_DB]) do
      its('output') { should eq 'INSERT 0 1' }
    end

    describe admin_sql.query('ALTER TABLE stig_test ADD COLUMN name text;', [PG_DB]) do
      its('output') { should eq 'ALTER TABLE' }
    end

    describe admin_sql.query('show pgaudit.log_catalog') do
      its('output') { should eq 'on' }
    end

    describe admin_sql.query('show pgaudit.log_level') do
      its('output') { should eq 'log' }
    end

    describe admin_sql.query('show pgaudit.log_parameter') do
      its('output') { should eq 'on' }
    end

    describe admin_sql.query('show pgaudit.log_statement_once') do
      its('output') { should eq 'off' }
    end

    describe admin_sql.query('show pgaudit.log') do
      its('output') { should eq 'ddl,read,role, write' }
    end

    describe admin_sql.query('CREATE ROLE foostigtest LOGIN CONNECTION LIMIT 100;') do
      its('output') { should eq 'CREATE ROLE' }
    end

    describe admin_sql.query('SET ROLE foostigtest; UPDATE stig_test SET id = 1 WHERE id = 0;', [PG_DB]) do
      its('output') { should match /ERROR:  permission denied for relation stig_test/ }
    end

   describe admin_sql.query('DROP TABLE stig_test;', [PG_DB]) do
      its('output') { should eq 'DROP TABLE' }
   end
  
  describe admin_sql.query('DROP ROLE foostigtest') do
      its('output') { should eq 'DROP ROLE' }
  end   

    describe postgres_conf(PG_CONF_FILE) do
      its('log_error_verbosity') { should eq 'default' }
      its('log_duration') { should eq 'on'}
    end
  end
